File: minesweeper.js
see: https://brandonjgoodman.com/minesweeper
- Hold ALT and click to mark a mine
- Click to reveal a square (can click on numbers too if the appropriate number of mines are marked)

Time:
  - BFS: O(M*N)
  - DFS: O(M*N)

Space:
  - BFS: O(M+N) maximum elements in the BFS would be the perimeter of our board
  - DFS: O(M*N)

/**
 * @param {character[][]} board
 * @param {number[]} click
 * @return {character[][]}
 */

// BFS
let dx;
let dy;
var updateBoard = function(board, click) {
    if (!board || !board.length) return board;
    
    const m = board.length;
    const n = board[0].length;
    const row = click[0];
    const col = click[1];
    
    if (board[click[0]][click[1]] == 'M') {
        board[click[0]][click[1]] = 'X';
        return board;
    }
    
    dx = [0, 0, 1, -1, 1, 1, -1, -1];
    dy = [1, -1, 0, 0, 1, -1, 1, -1];
    
    const q = [];
    q.push([row, col]);
    board[row][col] = 'B';
    
    while (q.length > 0) {
        const curr = q.shift(),
              row = curr[0],
              col = curr[1];
        const count = countMines(board, row, col);
        if (count == 0) {
            for(let i = 0; i < 8; i++) {
                let x = row + dx[i];
                let y = col + dy[i];

                if (isValid(board, x, y) && board[x][y] == 'E') {
                    board[x][y] = 'B';
                    q.push([x, y]);
                }
            }
        } else {
            board[row][col] = `${count}`;
        }
        
    }

    return board;
}


// DFS
// let dx;
// let dy;
// var updateBoard = function(board, click) {
//     if (!board || !board.length) return board;
    
//     const m = board.length;
//     const n = board[0].length;
//     const row = click[0];
//     const col = click[1];

//     if (board[click[0]][click[1]] == 'M') {
//         board[click[0]][click[1]] = 'X';
//         return board;
//     }
    
//     dx = [0, 0, 1, -1, 1, 1, -1, -1];
//     dy = [1, -1, 0, 0, 1, -1, 1, -1];
    
//     dfsVisit(board, row, col);

//     return board;
// };

// const dfsVisit = function(board, row, col) {
//     const count = countMines(board, row, col);
//     if (count == 0) {
//         board[row][col] = 'B';
//         for(let i = 0; i < 8; i++) {
//             let x = row + dx[i];
//             let y = col + dy[i];
                
//             if (isValid(board, x, y) && board[x][y] == 'E') {
//                 dfsVisit(board, x, y);
//             }
//         }
//     } else {
//         board[row][col] = `${count}`;
//     }
// }

const countMines = function(board, row, col) {
    let count = 0;
    for (let i = 0; i < 8; i++) {
        let x = row + dx[i];
        let y = col + dy[i];
        if (isValid(board, x, y) && board[x][y] == 'M') count++;
    }
    
    return count;
}

const isValid = function(board, x, y) {
    return x >= 0 && x < board.length && y >= 0 && y < board[0].length;
}

File: mineSweeper.py
class Solution:
    
    """
    Description: play minesweeper game, starting with a given click on a board 
    
    Time Complexity: O(m*n)
    Space Complexity: O(m*n)
    where, m and n are dimensions of the board
    
    Approach: BFS
    - if click is on mine, convert the value to "M" and return board
    - start with a queue having the indices of the click, convert to "B" (edge case)
    - pop the queue and check number of mines in all 8 directions:
      + if there are no mines around, search for "E" in all directions, revert the values to "B"
      + else, revert values to number of mines (use a separate function to count number of mines)
    """

    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        
        from collections import deque
        self.m = len(board); self.n = len(board[0])
        ir, ic = click

        if board[ir][ic] == "M": 
            board[ir][ic] = "X"
            return board
        
        # directions to search, diagonal included
        self.dirs = [(-1,-1), (0,-1), (1,-1), (1,0), (1,1), (0,1), (-1,1), (-1,0)]
        
        queue = deque()
        queue.append((ir, ic))
        board[ir][ic] = "B"
        
        while queue:
            
            curr = queue.popleft()
            n_mines = self.count_mines(board, curr[0], curr[1])
            if n_mines == 0:
                for d in self.dirs:
                    row = curr[0] + d[0]
                    col = curr[1] + d[1]
                    if row >= 0 and col >= 0 and row < self.m and col < self.n and board[row][col] == "E":
                        board[row][col] = "B"
                        queue.append((row, col))
            else:
                board[curr[0]][curr[1]] = str(n_mines)              
                
        return board

    def count_mines(self, board, ir, ic):
        n_mines = 0
        for d in self.dirs:
            row = ir + d[0]
            col = ic + d[1]
            if row >= 0 and col >= 0 and row < self.m and col < self.n and board[row][col] == "M": 
                n_mines += 1

        return n_mines  
